mountpoint-s3/tests/fuse_tests/readdir_test.rs

```diff
@@ -30,38 +33,107 @@ fn prepare_fs(mut put_object_fn: PutObjectFn, map: &HashMap<String, File>) {
+    let dirs: Vec<_> = dir.map(|f| f.unwrap()).collect();
+    assert_eq!(
+        dirs.iter()
+            .map(|f| f.path().file_name().unwrap().to_str().unwrap().to_owned())
+            .collect::<Vec<_>>(),
+        expected_list
+    );
+}
+
+fn readdir_while_writing<F>(creator_fn: F, prefix: &str, new_files: Vec<&str>)
+where
+    F: FnOnce(&str, S3FilesystemConfig) -> (TempDir, BackgroundSession, PutObjectFn),
+{
+    let readdir_size = 5;
+    let filesystem_config = S3FilesystemConfig {
+        readdir_size,
+        ..Default::default()
+    };
 
-    for f in files.drain(..) {
-        let path = f.path();
-        let name = path.file_name().unwrap().to_str().unwrap();
-        assert!(map.remove(name).is_some());
+    let mut map = HashMap::new();
+    let mut expected_list = Vec::new();
+    for i in 0..readdir_size * 4 {
+        let file_name = format!("file{i}");
+        map.insert(file_name.clone(), File::new((i % 256) as u8, 10 * i));
+        expected_list.push(file_name);
     }
 
-    println!("... map is {map:?}");
-    assert!(map.is_empty());
+    let (mount_point, _session, put_object_fn) = creator_fn(prefix, filesystem_config);
+
+    prepare_fs(put_object_fn, &map);
+
+    const OBJECT_SIZE: usize = 1024;
+    let mut opened_files = Vec::new();
+    for file_name in new_files {
+        let path = mount_point.path().join(file_name);
+        let mut options = fs::File::options();
+        options.write(true);
+        options.create(true);
+        let f = options.open(path).unwrap();
+
+        let mut rng = ChaCha20Rng::seed_from_u64(0x12345678 + OBJECT_SIZE as u64);
+        let mut body = vec![0u8; OBJECT_SIZE];
+        rng.fill(&mut body[..]);
+        opened_files.push(f);
+        map.insert(file_name.to_owned(), File::new(0, 0));
+        expected_list.push(file_name.to_owned());
+    }
+    expected_list.sort();
+
+    let dir = fs::read_dir(mount_point.path()).unwrap();
+    let dirs: Vec<_> = dir.map(|f| f.unwrap()).collect();
+    assert_eq!(
+        dirs.iter()
+            .map(|f| f.path().file_name().unwrap().to_str().unwrap().to_owned())
+            .collect::<Vec<_>>(),
+        expected_list
+    );
 }
 
+#[cfg(feature = "s3_tests")]
 #[test]
-fn readdir_1_mock() {
-    readdir_1(crate::fuse_tests::mock_session::new, "");
+fn readdir_s3() {
+    readdir(crate::fuse_tests::s3_session::new, "readdir_s3");
+}
+
+#[cfg(feature = "s3_tests")]
+#[test_case(vec!["aaa", "file150", "file151", "zzz"]; "new files at various positions")]
+fn readdir_while_writing_s3(new_files: Vec<&str>) {
+    readdir_while_writing(crate::fuse_tests::s3_session::new, "readdir_s3", new_files);
+}
+
+#[test]
+fn readdir_mock() {
+    readdir(crate::fuse_tests::mock_session::new, "");
+}
+
+#[test_case(vec!["aaa", "file150", "file151", "zzz"]; "new files at various positions")]
+fn readdir_while_writing_mock(new_files: Vec<&str>) {
+    readdir_while_writing(crate::fuse_tests::mock_session::new, "readdir_s3", new_files);
```

Comment:
This behavior might be worth writing either a proptest or a randomized test for. The interaction between the ListObjects pagination and which objects are being written is pretty subtle with a few corner cases, so these testing approaches might give us more confidence we're covering the corner cases appropriately.

Reply:
Agree that we will need a proptest for this, and I found another problem reported in https://github.com/awslabs/mountpoint-s3/issues/190, it seems like our current pagination logic might not be correct?

Reply:
These tests are now using randomized file names.